src/utils/cosmosdb_model_versioned_ttl.ts

```diff
@@ -0,0 +1,234 @@
+const BatchResult = t.readonlyArray(t.type({ statusCode: NonNegativeInteger }));
+
+/**
+  The purpose of this class is to isolate the `updateTTL` logic, 
+  it should be used only for those models which does not generate a lot 
+  of versions because the logic to update the ttl is quite expensive.
+ 
+  NB: 
+  if You have more than 100 versions the update operation could not be transactional, 
+  see https://learn.microsoft.com/en-us/azure/cosmos-db/nosql/transactional-batch?tabs=dotnet
+ */
+
+export const RetrievedVersionedModelTTL = t.intersection([
+  CosmosResourceTTL,
+  VersionedModel
+]);
+
+export type RetrievedVersionedModelTTL = t.TypeOf<
+  typeof RetrievedVersionedModelTTL
+>;
+
+export class CosmosdbModelVersionedTTL<
+  T,
+  TN extends Readonly<T>,
+  TR extends Readonly<T & RetrievedVersionedModelTTL>,
+  ModelIdKey extends keyof T,
+  PartitionKey extends keyof T = ModelIdKey
+> extends CosmosdbModelVersioned<T, TN, TR, ModelIdKey, PartitionKey> {
+  constructor(
+    container: Container,
+    protected readonly newVersionedItemT: t.Type<TN, ItemDefinition, unknown>,
+    protected readonly retrievedItemT: t.Type<TR, unknown, unknown>,
+    protected readonly modelIdKey: ModelIdKey,
+    protected readonly partitionKey?: PartitionKey
+  ) {
+    super(
+      container,
+      newVersionedItemT,
+      retrievedItemT,
+      modelIdKey,
+      partitionKey
+    );
+  }
+
+  /**
+   * Using transactionalBatch updates the ttl field for all the versions
+   * of the document identified by the searchKey.
+   *
+   * @param searchKey
+   * @param ttl
+   * @returns Either a Cosmos Error or the number of updated elements
+   */
+  public updateTTLForAllVersions(
+    searchKey: DocumentSearchKey<TR, ModelIdKey, PartitionKey>,
+    ttl: RetrievedVersionedModelTTL["ttl"]
+  ): TE.TaskEither<CosmosErrors, number> {
+    const partitionKey = ((searchKey.length === 1
+      ? searchKey[0]
+      : searchKey[1]) as unknown) as TN[PartitionKey];
+
+    return pipe(
+      this.findAllVersionsBySearchKey(searchKey),
+      TE.map(RA.rights),
+      TE.map(
+        RA.map(doc => ({
+          id: doc.id,
+          operationType: "Patch" as const,
+          resourceBody: {
+            operations: [
+              {
+                // `add` will create or update the path with the passed value
+                op: PatchOperationType.add,
+                path: `/ttl`,
+                value: ttl
+              }
+            ]
+          }
+        }))
+      ),
+      TE.map(RA.chunksOf(100)),
+      TE.map(
+        RA.map(chunk =>
+          pipe(
+            chunk,
+            RA.toArray,
+            operations => this.batch(operations, partitionKey),
+            TE.chainW(
+              TE.fromPredicate(
+                response => response.code === 200,
+                _ => {
+                  const firstChunkId = chunk[0]?.id ?? -1;
+                  const lastChunkId = chunk[chunk.length - 1]?.id ?? -1;
```

Comment:
A empty chunk don't make sense in here. We are executing a batch without any operation? Skip the batch operation if chunk is an empty array instead.
```suggestion
                  const firstChunkId = chunk[0].id;
                  const lastChunkId = chunk[chunk.length - 1].id;
```

Reply:
Fine, but we need to make sure that `findAllVersionBySearchKey` did not return an empty array.